### PR TITLE
libs/libc: Breakdown LIBC_BUILD_STRING into specific string operation…

### PR DESCRIPTION
**Pull request by yddhotnet**
## Summary

Provide a way to only customize specific string operations, such as for memcpy with the DMA capability by ROM.

### CODE REVIEW

**Comment by mu578** [comment]
If you do so, i.e per string functions, they might be more, you never know, it should be propagated to every existing target then. Or then should be something inside `memcpy` with another bypass flag, make a choice here: either it is a new way, a generalization either specific to only few on demand.

File: libs/libc/string/lib_vikmemcpy.c [hunk 1]
```diff
@@ -64,7 +64,7 @@
 
 #include "libc.h"
 
-#if !defined(CONFIG_LIBC_ARCH_MEMCPY) && defined(LIBC_BUILD_STRING)
+#if !defined(CONFIG_LIBC_ARCH_MEMCPY) && defined(LIBC_BUILD_VIKMEMCPY)
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_vikmemcpy.c:67]:
```suggestion
#if !defined(CONFIG_LIBC_ARCH_MEMCPY) && defined(LIBC_BUILD_MEMCPY)
```

File: libs/libc/machine/Kconfig [hunk 2]
```diff
@@ -137,12 +137,179 @@ config LIBC_ARCH_ELF_64BIT
 	default n
 	depends on LIBC_ARCH_ELF
 
-config LIBC_PREVENT_STRING_KERNEL
```
Comment (xiaoxiang781216) [inline libs/libc/machine/Kconfig:140]: let's keep LIBC_PREVENT_STRING_KERNEL and select LIBC_PREVENT_*_KERNEL

File: libs/libc/machine/Kconfig [hunk 3]
```diff
@@ -137,12 +137,179 @@ config LIBC_ARCH_ELF_64BIT
...
+config LIBC_PREVENT_STRRCHR_KERNEL
```
Comment (xiaoxiang781216) [inline libs/libc/machine/Kconfig:140]: let's sort the new option

File: libs/libc/machine/Kconfig [hunk 4]
```diff
@@ -151,15 +318,325 @@ config LIBC_PREVENT_STRING_KERNEL
 		NuttX's software-defined version of the libc or arch-specific
 		assembly version is built instead.
 
-config LIBC_PREVENT_STRING_USER
```
Comment (xiaoxiang781216) [inline libs/libc/machine/Kconfig:154]: let's keep LIBC_PREVENT_STRING_USER and select LIBC_PREVENT_*_USER

File: libs/libc/libc.h [hunk 5]
```diff
@@ -58,7 +58,27 @@
 #if defined(CONFIG_BUILD_FLAT) || \
     ((!defined(CONFIG_LIBC_PREVENT_STRING_USER) && !defined(__KERNEL__))  || \
      (!defined(CONFIG_LIBC_PREVENT_STRING_KERNEL) && defined(__KERNEL__)))
-#  define LIBC_BUILD_STRING
```
Comment (xiaoxiang781216) [inline libs/libc/libc.h:61]: change individual option like this:
```
#if defined(CONFIG_BUILD_FLAT) || \
    ((!defined(CONFIG_LIBC_PREVENT_STRRCHR_USER) && !defined(__KERNEL__))  || \
     (!defined(CONFIG_LIBC_PREVENT_STRRCHR_KERNEL) && defined(__KERNEL__)))
#  define LIBC_BUILD_STRRCHR
#endif
```

**Comment by xiaoxiang781216** [comment]
> If you do so, i.e per string functions, they might be more, you never know, it should be propagated to every existing target then. Or then should be something inside `memcpy` with another bypass flag, make a choice here: either it is a new way, a generalization either specific to only few on demand. Personal opinion: looks like bad juju in that shape.

this can be fixed by let LIBC_PREVENT_STRING_[USER|KERNEL] select LIBC_PREVENT_XXX_[USER|KERNEL].

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@tmedicci could you review this patch?

File: libs/libc/libc.h (hunk 5 above)
Comment (yddhotnet) [reply libs/libc/libc.h:61]: Thanks. Updated as comments.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**